## Random Forest Decision Tree

Training Data Score: 0.99545159194282

Testing Data Score: 0.2573099415204678


In [1]:
# Dependencies
import pandas as pd
from sklearn.preprocessing import LabelEncoder
import numpy as np
from sklearn import tree
import os

In [2]:
# import processed data

path = "data/"
file = "Best_disposition_data.csv"
path_file = path + file

df = pd.read_csv(path_file)
df = df.drop('Unnamed: 0', axis=1)
df

DATE_year  DATE_month  DATE_week  DATE_day  Hour  Minute  Temp(K)  \
0          2012          10         40         4    20       0   293.31   
1          2012          10         40         4    23       0   291.82   
2          2012          10         40         5     0       1   289.74   
3          2012          10         41        10    14       0   281.87   
4          2012          10         42        15    12      30   292.15   
...         ...         ...        ...       ...   ...     ...      ...   
2047       2017           8         32        11    13       0   293.42   
2048       2017           8         33        20    20      29   304.17   
2049       2017           9         39        27    17      30   297.76   
2050       2017          10         41        13    16      30   289.81   
2051       2017          11         45         9    18      30   282.82   

      Humidity%  Pressure(atm)  wind  Description  disposition  
0          56.0         1017.0   7.0            0            0  
1          68.0         1017.0   3.0           13            2  
2          63.0         1018.0   3.0           13            0  
3          66.0         1020.0   2.0            2            2  
4          82.0         1009.0   8.0            7            0  
...         ...            ...   ...          ...          ...  
2047       88.0         1017.0   2.0           14            4  
2048       54.0         1018.0   1.0           14            2  
2049       73.0         1018.0   5.0            0            2  
2050       67.0         1023.0   2.0           13            2  
2051       70.0         1020.0   5.0           14            2  

[2052 rows x 12 columns]

In [3]:
# Create dataframe for running decision tree
# This was done to upload existing code with similar variables
# Main Data set: data_df

data_df = df.drop("disposition", axis=1)
target = df["disposition"]
#target


feature_names = data_df.columns
data_df.head()

DATE_year  DATE_month  DATE_week  DATE_day  Hour  Minute  Temp(K)  \
0       2012          10         40         4    20       0   293.31   
1       2012          10         40         4    23       0   291.82   
2       2012          10         40         5     0       1   289.74   
3       2012          10         41        10    14       0   281.87   
4       2012          10         42        15    12      30   292.15   

   Humidity%  Pressure(atm)  wind  Description  
0       56.0         1017.0   7.0            0  
1       68.0         1017.0   3.0           13  
2       63.0         1018.0   3.0           13  
3       66.0         1020.0   2.0            2  
4       82.0         1009.0   8.0            7

In [4]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(data_df, target, random_state=42)

In [5]:
from sklearn.ensemble import RandomForestClassifier
rf = RandomForestClassifier(n_estimators=200)
rf = rf.fit(X_train, y_train)
rf.score(X_test, y_test)

0.2553606237816764

In [6]:
print('Test Acc: %.3f' % rf.score(X_test, y_test))

Test Acc: 0.255


In [7]:
print(f"Training Data Score: {rf.score(X_train, y_train)}")
print(f"Testing Data Score: {rf.score(X_test, y_test)}")

Training Data Score: 0.99545159194282
Testing Data Score: 0.2553606237816764


## Integrate new encoded data
### label asignments
Districts

DT:0

ND:1

NE:2

NW:3

SE:4

SW:5

Bike Value

Street address -----Will not be used Experiment value

In [8]:
# Make predictions with the hypertuned model
predictions = rf.predict(X_test)

# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions,
                            target_names=["DT", "ND", "NE","NW","SE","SW"]))

              precision    recall  f1-score   support

          DT       0.32      0.58      0.41       121
          ND       0.24      0.12      0.16        81
          NE       0.25      0.36      0.29       104
          NW       0.08      0.03      0.05        62
          SE       0.20      0.09      0.12        81
          SW       0.13      0.08      0.10        64

    accuracy                           0.26       513
   macro avg       0.20      0.21      0.19       513
weighted avg       0.22      0.26      0.22       513



In [9]:
sorted(zip(rf.feature_importances_, feature_names), reverse=True)

[(0.13387692509113028, 'Temp(K)'),
 (0.11754797933241284, 'Humidity%'),
 (0.10896881162688919, 'DATE_day'),
 (0.10827062485125462, 'Pressure(atm)'),
 (0.10212941197276003, 'Hour'),
 (0.09115915913978336, 'DATE_week'),
 (0.08977677315717494, 'Minute'),
 (0.07100540745177368, 'Description'),
 (0.0699921109387485, 'wind'),
 (0.05415174930540183, 'DATE_month'),
 (0.05312104713267085, 'DATE_year')]

## Making Predictions with new data
<b>Label asignments</b>

**DT:0

**NE:2

**NW:3

**SE:4

**SW:5

DATE_year	DATE_month	DATE_week	DATE_day	Hour	Minute	Temp(K)	Humidity%	Pressure(atm)	wind	Street (Coded)
2012	    10	        40	        4	        20	    0	    293.31	      56	        1017	  7	     1675	

2012	    12	        49	        7	        23	    0	    281.34	      70	        1016	  2	     1620	

2017	     7	        28	        6	        13	   27	    292.06	      88	        1017	  1	     618	


SET 1 ANSWER 0

SET 2 ANSWER 2

SET 3 ANSER 0



In [10]:
import numpy as np
new_data =   np.array([[2012,12, 49, 7, 23, 0, 281.34, 70, 1016, 2]])
new_data_2 = np.array([[2017,7, 28, 6, 13, 27, 292.06, 88, 1017, 1]]) 

In [11]:
# quick test
rf.predict(new_data)

ValueError: Number of features of the model must match the input. Model n_features is 11 and input n_features is 10 

In [12]:
# Test the model

print(f"Predicted class: {rf.predict(new_data)}")

ValueError: Number of features of the model must match the input. Model n_features is 11 and input n_features is 10 

In [15]:
 # save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
#filename = 'z1_randomForest_model.sav'
#joblib.dump(rf, filename)

Use data set from scaleMaxMin

In [16]:
# import processed data that was scaled using minmaxscale

path = "data/"
file = "minmaxscaled_disposition_data.csv"
path_file = path + file

scaled_df = pd.read_csv(path_file)
scaled_df = scaled_df.drop('Unnamed: 0', axis=1)
scaled_df

DATE_year  DATE_month  DATE_week  DATE_day      Hour    Minute  \
0           0.0    0.818182   0.750000  0.100000  0.869565  0.000000   
1           0.0    0.818182   0.750000  0.100000  1.000000  0.000000   
2           0.0    0.818182   0.750000  0.133333  0.000000  0.016949   
3           0.0    0.818182   0.769231  0.300000  0.608696  0.000000   
4           0.0    0.818182   0.788462  0.466667  0.521739  0.508475   
...         ...         ...        ...       ...       ...       ...   
2047        1.0    0.636364   0.596154  0.333333  0.565217  0.000000   
2048        1.0    0.636364   0.615385  0.633333  0.869565  0.491525   
2049        1.0    0.727273   0.730769  0.866667  0.739130  0.508475   
2050        1.0    0.818182   0.769231  0.400000  0.695652  0.508475   
2051        1.0    0.909091   0.846154  0.266667  0.782609  0.508475   

       Temp(K)  Humidity%  Pressure(atm)      wind  disposition  
0     0.704366     0.4500       0.289474  0.538462            0  
1     0.675707     0.6000       0.289474  0.230769            2  
2     0.635699     0.5375       0.298246  0.230769            0  
3     0.484324     0.5750       0.315789  0.153846            2  
4     0.682054     0.7750       0.219298  0.615385            0  
...        ...        ...            ...       ...          ...  
2047  0.706482     0.8500       0.289474  0.153846            4  
2048  0.913253     0.4250       0.298246  0.076923            2  
2049  0.789960     0.6625       0.298246  0.384615            2  
2050  0.637046     0.5875       0.342105  0.153846            2  
2051  0.502597     0.6250       0.315789  0.384615            2  

[2052 rows x 11 columns]

In [17]:
# Create dataframe for running decision tree
# This was done to upload existing code with similar variables
# Main Data set: data_df

scaled_data_df = scaled_df.drop("disposition", axis=1)
target2 = scaled_df["disposition"]
#target


feature_names = scaled_data_df.columns
scaled_data_df.head()

DATE_year  DATE_month  DATE_week  DATE_day      Hour    Minute   Temp(K)  \
0        0.0    0.818182   0.750000  0.100000  0.869565  0.000000  0.704366   
1        0.0    0.818182   0.750000  0.100000  1.000000  0.000000  0.675707   
2        0.0    0.818182   0.750000  0.133333  0.000000  0.016949  0.635699   
3        0.0    0.818182   0.769231  0.300000  0.608696  0.000000  0.484324   
4        0.0    0.818182   0.788462  0.466667  0.521739  0.508475  0.682054   

   Humidity%  Pressure(atm)      wind  
0     0.4500       0.289474  0.538462  
1     0.6000       0.289474  0.230769  
2     0.5375       0.298246  0.230769  
3     0.5750       0.315789  0.153846  
4     0.7750       0.219298  0.615385

In [18]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(scaled_data_df, target2, random_state=42)

In [19]:
rfs = RandomForestClassifier(n_estimators=200)
rfs = rf.fit(X_train, y_train)
rfs.score(X_test, y_test)

0.2573099415204678

In [20]:
print(f"Training Data Score: {rfs.score(X_train, y_train)}")
print(f"Testing Data Score: {rfs.score(X_test, y_test)}")

Training Data Score: 0.99545159194282
Testing Data Score: 0.2573099415204678
